common/cache.go

```diff
@@ -0,0 +1,100 @@
+package common
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/zcash-hackworks/lightwalletd/walletrpc"
+)
+
+type BlockCacheEntry struct {
+	data []byte
+	hash []byte
+}
+
+type BlockCache struct {
+	MaxEntries int
+
+	FirstBlock int
+	LastBlock  int
+
+	m map[int]*BlockCacheEntry
+
+	mutex sync.RWMutex
+}
+
+func NewBlockCache(maxEntries int) *BlockCache {
+	return &BlockCache{
+		MaxEntries: maxEntries,
+		FirstBlock: -1,
+		LastBlock:  -1,
+		m:          make(map[int]*BlockCacheEntry),
+	}
+}
+
+func (c *BlockCache) Add(height int, block *walletrpc.CompactBlock) (error, bool) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// If we already have this block or any higher blocks, a reorg
+	// must have occurred; these must be re-added
+	for i := height; i <= c.LastBlock; i++ {
+		delete(c.m, i)
+	}
```

Comment:
Nice! I'm convinced this new approach ensures there's never any reorged-away blocks in the cache. A couple issues remain:

If there's a reorg detected or the JSON unmarshalling fails below, then the function will return after this loop has deleted from `height` up to `c.LastBlock`, but `c.LastBlock` won't have been updated (leaving nil blocks in the cache). The same is also true of `c.FirstBlock`, calling this with a really low height will empty the cache but `c.FirstBlock` won't be updated if the function exits through those returns. We should make sure the cache is in a valid state immediately after this loop, which probably means something like this:

```
if height <= c.FirstBlock {
    // the entire cache was just emptied
    c.FirstBlock = -1;
    c.LastBlock = -1;
    // maybe assert that c.m is really empty here just in case
} else {
    // height > c.FirstBlock, so we know there's still a block at height - 1 in the cache
    c.LastBlock = height - 1;
}
```

(Please double-check this is correct since I haven't fully thought it through)

It's also still possible to add blocks at `height` > `c.LastBlock` + `1` to get the cache into a state containing `nil` blocks (I think this case just needs to be explicitly checked for).